### ISSUE TITLE
CUDA device selection not working using CUDA_VISIBLE_DEVICES

### ISSUE THREAD

**Issue by dittothat** [MIC-DKFZ/nnUNet#1798]
I would like to train in parallel on two GPUs, but after starting on device 0 using:
`nnUNetv2_train 2 2d 2`
Trying to start on device 1 using:
`CUDA_VISIBLE_DEVICES=1 nnUNetv2_train 2 2d 3`
Reverts to:
`Using device: cuda:0`
Is this expected behavior?
Thank you.

If relevant:
NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2 
acvl-utils==0.2
astroid==3.0.1
batchgenerators==0.25
certifi==2023.7.22
charset-normalizer==3.3.1
connected-components-3d==3.12.3
contourpy==1.1.1
cuda-python==11.7.0
cycler==0.12.1
Cython==3.0.5
dicom2nifti==2.4.8
dill==0.3.7
dynamic-network-architectures==0.2
filelock==3.13.1
fonttools==4.43.1
fsspec==2023.10.0
future==0.18.3
graphviz==0.20.1
idna==3.4
imagecodecs==2023.9.18
imageio==2.31.6
isort==5.12.0
Jinja2==3.1.2
joblib==1.3.2
kiwisolver==1.4.5
lazy_loader==0.3
linecache2==1.0.0
MarkupSafe==2.1.3
matplotlib==3.8.1
mccabe==0.7.0
mpmath==1.3.0
networkx==3.2.1
nibabel==5.1.0
nnunetv2==2.2.1
nptyping==2.5.0
numpy==1.26.1
nvidia-cublas-cu12==12.1.3.1
nvidia-cuda-cupti-cu12==12.1.105
nvidia-cuda-nvrtc-cu12==12.1.105
nvidia-cuda-runtime-cu12==12.1.105
nvidia-cudnn-cu12==8.9.2.26
nvidia-cufft-cu12==11.0.2.54
nvidia-curand-cu12==10.3.2.106
nvidia-cusolver-cu12==11.4.5.107
nvidia-cusparse-cu12==12.1.0.106
nvidia-nccl-cu12==2.18.1
nvidia-nvjitlink-cu12==12.3.52
nvidia-nvtx-cu12==12.1.105
packaging==23.2
pandas==2.1.2
Pillow==10.0.1
platformdirs==3.11.0
pydicom==2.4.3
pylint==3.0.2
pynrrd==1.0.0
pyparsing==3.1.1
python-dateutil==2.8.2
python-gdcm==3.0.22
python-toolbox==1.0.11
pytz==2023.3.post1
PyYAML==6.0.1
requests==2.31.0
scikit-image==0.22.0
scikit-learn==1.3.2
scipy==1.11.3
seaborn==0.13.0
SimpleITK==2.3.0
six==1.16.0
slicerio==0.1.8
sympy==1.12
threadpoolctl==3.2.0
tifffile==2023.9.26
tomli==2.0.1
tomlkit==0.12.2
torch==2.1.0
torchaudio==2.1.0
torchvision==0.16.0
tqdm==4.66.1
traceback2==1.4.0
tract_querier==0.1
triton==2.1.0
typing_extensions==4.8.0
tzdata==2023.3
unittest2==1.1.0
urllib3==2.0.7
yacs==0.1.8

**Comment by dittothat** [comment]
So I investigated some more.
I halted all training and then tried:
```
CUDA_VISIBLE_DEVICES=0 nnUNetv2_train --c 2 2d 2
CUDA_VISIBLE_DEVICES=1 nnUNetv2_train --c 2 2d 3
```
And got the same behavior in that both processes reported using `cuda:0`.
But looking at the output of nvivdia-smi the two processes loaded on each GPU:
```
Thu Nov 16 14:26:23 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA TITAN V                 Off | 00000000:AF:00.0 Off |                  N/A |
| 59%   82C    P2              84W / 250W |   4440MiB / 12288MiB |     26%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA TITAN V                 Off | 00000000:D8:00.0 Off |                  N/A |
| 40%   59C    P2              47W / 250W |   4598MiB / 12288MiB |      1%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   1626704      C   ...b/research/segment/venv/bin/python3     4436MiB |
|    1   N/A  N/A   1632192      C   ...b/research/segment/venv/bin/python3     4594MiB |
+---------------------------------------------------------------------------------------+
```
So it must have been something with my initial calls coupled with the misleading ```cuda:0``` reporting.

**Comment by ancestor-mithril** [comment]
`cuda:0` does not refer to GPU:0. `cuda:0` refers to the first visible GPU for the current process.
